### PR TITLE
assumeutxo: keep cache when flushing snapshot (#17487 followup)

### PR DESCRIPTION
**Pull request by jamesob**
Make use of #17487.

Benchmark results available here:
  https://github.com/bitcoin/bitcoin/pull/17487#issuecomment-557226923

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -5033,15 +5033,19 @@ bool ChainstateManager::ActivateSnapshot(
     return true;
 }
 
-static void FlushSnapshotToDisk(CCoinsViewCache& coins_cache, bool snapshot_loaded)
+static void FlushSnapshotToDisk(CCoinsViewCache& coins_cache, bool snapshot_loaded, bool erase=true)
 {
     LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE(
         strprintf("%s (%.2f MB)",
                   snapshot_loaded ? "saving snapshot chainstate" : "flushing coins cache",
```
Comment (Sjors) [inline src/validation.cpp:5040]: nit: `(erase ? "syncing" : "flushing") + "coins cache"`

**Review by Sjors — APPROVED** [review]
utACK d6f54dac8dee776188fd88d725e2728e0822b33f

Based on testing #17487 I think it's safe. I'd like to take this for a spin on a rebase of #15606 (the bits that enable `loadtxoutset`), to see if I can reproduce the performance improvement. But imo that doesn't have to hold back this PR.

**Comment by jamesob** [comment]
Thanks for the look @Sjors. I also realized that there were two outlier calls where we weren't asserting that Flush()/Sync() has succeeded, so I've fixed those.

File: src/validation.cpp [hunk 2]
```diff
@@ -5036,12 +5036,12 @@ bool ChainstateManager::ActivateSnapshot(
 static void FlushSnapshotToDisk(CCoinsViewCache& coins_cache, bool snapshot_loaded)
 {
     LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE(
-        strprintf("%s (%.2f MB)",
-                  snapshot_loaded ? "saving snapshot chainstate" : "flushing coins cache",
+        strprintf("%s coins cache (%.2f MB)",
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/validation.cpp:5041]: Sounds good, fixed.

**Review by jamesob — COMMENTED** [review]
Thanks for the feedback @Sjors.

**Comment by Sjors** [comment]
re-utACK 937016b1917a0a07fa579c96fc797237f5027cd1

File: src/validation.cpp [hunk 3]
```diff
@@ -5033,15 +5035,19 @@ bool ChainstateManager::ActivateSnapshot(
...
         strprintf("[snapshot] %s coins cache (%.2f MB)",
                   (snapshot_loaded ? "syncing" : "flushing"),
                   coins_cache.DynamicMemoryUsage() / (1000 * 1000)),
         BCLog::LogFlags::ALL);
 
-    snapshot_loaded ? coins_cache.Sync() : coins_cache.Flush();
+    if (snapshot_loaded ? coins_cache.Sync() : coins_cache.Flush()) {
+        return true;
+    } else {
+        return AbortNode("Failed to write to coin database");
```
Comment (Sjors) [inline src/validation.cpp:5049]: Same question about `AbortNode` vs. `error` though.

Comment (jamesob) [reply src/validation.cpp:5049]: `AbortNode()` is used elsewhere in the function that this function is called from, although I did change another reference that you pointed out what inconsistent.

**Comment by Sjors** [comment]
From my testing of #15606 - without this PR - I get the impression that flushing isn't working, specifically that RAM is not freed up. It could just be confusion on my end from rebase hell, but I'd like to make sure we're not obscuring a bug by merging this.

**Comment by jamesob** [comment]
@Sjors how did you detect that RAM *wasn't* being freed? Did you verify that the flush did not happen based on an absence of logs? 

Measures of memory can be deceptive; I remember bitcoinperf spuriously detecting an increase in memory usage because once claimed, the process' resident set size (RSS) doesn't necessarily go down upon free.

**Comment by jamesob** [comment]
Actually this _is_ broken at the moment, since although we don't flush after snapshot load, we do flush (and not sync) in the MaybeRebalanceCaches() -> FlushStateToDisk() call right after activating the chainstate snapshot. Will close this and think about how to rework it.

**Pull request closed without merging**